### PR TITLE
Merge the demo app with the main package

### LINKED ISSUE #26: Merge the demo app with the main package
Having a separate package for the demo app is inconsistent with all the other stuff we see in the npm/JS world. I suggest you put the demo codebase into a `src/demo` folder, put it into `.npmignore` and have the `yarn start` command serve the demo from the main package. All the extra dependencies can go as development ones.

### PR DESCRIPTION
**Pull request by brumik**
The demo application's package.json is merged with the root, so there is no need to install duplicate packages or switch directories when running the demo app. Also, it now uses the webpack-dev-server to serve the content, not the react-scripts package. The demo-app folder is renamed to demo to go with the convention.

Fixes #26

### CODE REVIEW

**Comment by brumik** [comment]
@skateman @Hyperkid123 I added NODE_ENV for the react-script as is described [here](https://facebook.github.io/create-react-app/docs/advanced-configuration) so the demo app uses the root node modules. Is it okay?

File: package.json [hunk 1]
```diff
@@ -26,34 +26,41 @@
     "react-dom": "^16.8.6"
   },
   "scripts": {
-    "start": "cd demo-app && react-scripts-ts start",
+    "start": "webpack-dev-server --mode development",
```
Comment (skateman) [inline package.json]: This is starting to look like what I wanted, nice! :+1:

File: package.json [hunk 2]
```diff
@@ -26,34 +26,41 @@
...
     "redux": "^4.0.1",
     "serve": "^11.0.0",
+    "style-loader": "^0.23.1",
+    "ts-loader": "^6.0.0",
     "tslint-loader": "^3.6.0",
     "typedoc": "^0.14.2",
     "typescript": "^3.4.5",
-    "webpack": "^4.30.0",
-    "webpack-cli": "^3.3.2"
+    "webpack": "^4.31.0",
```
Comment (skateman) [inline package.json]: In general I'm not sure it is a great idea to upgrade package versions in this PR. When you're doing something that's not dependent on version upgrades, you should not upgrade. The amount of changes should be always lowest as possible.

File: webpack.config.js [hunk 3]
```diff
@@ -26,15 +33,32 @@ module.exports = {
                 exclude: /node_modules/
             },
             {
-                test: /\.(ts|tsx)$/,
-                loader: 'awesome-typescript-loader',
-                exclude: [/node_modules/]
+                test: /\.(js|jsx)$/,
```
Comment (skateman) [inline webpack.config.js]: Please explain this change...

**Comment by skateman** [comment]
You still have a merge conflict, also the commits are not really making sense and the PR description even less.

**Comment by Hyperkid123** [comment]
@brumik here is a patch for the demo:

Some notes
- no need for externals in dev env
- webpack will server the js from its server. So the script reference must be relative to server not to your computer storage
- using the html plugin will add any js/style references automatically for you and it always matches your webpack config
- any other than default dev server setup is overkill(in most cases). So i deleted it

```diff
diff --git a/demo/public/index.html b/demo/public/index.html
index 5250d95..fa2507d 100644
--- a/demo/public/index.html
+++ b/demo/public/index.html
@@ -8,6 +8,5 @@
   </head>
   <body>
     <div id="root"></div>
-    <script src="dist/demo.js"></script>
   </body>
 </html>
diff --git a/demo/webpack.config.js b/demo/webpack.config.js
index 67ce041..417565a 100644
--- a/demo/webpack.config.js
+++ b/demo/webpack.config.js
@@ -1,5 +1,11 @@
 const { resolve } = require('path');
 const CopyWebpackPlugin = require('copy-webpack-plugin');
+const HtmlWebPackPlugin = require('html-webpack-plugin');
+
+const htmlPlugin = new HtmlWebPackPlugin({
+    template: './public/index.html',
+    filename: './index.html',
+});
 
 module.exports = {
     context: __dirname,
@@ -11,14 +17,9 @@ module.exports = {
         path: resolve(__dirname, './dist'),
         filename: '[name].js'
     },
-    devServer: {
-        publicPath: '/dist/',
-        contentBase: resolve(__dirname, './public'),
-        index: resolve(__dirname, './public/index.html'),
-        hot: true
-    },
     plugins: [
-        new CopyWebpackPlugin(['src/**/*.d.ts'])
+        new CopyWebpackPlugin(['src/**/*.d.ts']),
+        htmlPlugin
     ],
     module: {
         rules: [
@@ -58,12 +59,4 @@ module.exports = {
         ]
     },
     resolve: {extensions: ['.js', '.ts','.tsx', '.css']},
-    externals: {
-        'react': {
-            commonjs: 'react',
-            commonjs2: 'react',
-            amd: 'react',
-            root: 'React'
-        }
-    }
 };
diff --git a/package.json b/package.json
index 6cebdd2..3f767ca 100644
--- a/package.json
+++ b/package.json
@@ -47,6 +47,7 @@
     "css-loader": "^2.1.1",
     "file-loader": "^3.0.1",
     "font-awesome": "^4.7.0",
+    "html-webpack-plugin": "^3.2.0",
     "react-redux": "^7.0.3",
     "react-scripts-ts": "^3.1.0",
     "react-test-renderer": "^16.8.6",

```

File: webpack.config.js (hunk 3 above)
Comment (brumik) [reply webpack.config.js]: It was a mistake, just changed the typescript loader

**Comment by brumik** [comment]
@skateman Does it seems ok now?
@Hyperkid123 Thanks

**Comment by skateman** [comment]
> also the commits are not really making sense and the PR title & description even less

**Pull request merged**